Data dictionary UI for specifying allowed values for select types

kmtracey opened the pull request:
- add model for storing the allowed values and their descriptions
- fetch existing allowed values when initial page data is fetched
- display allowed values and an edit button when appropriate
- edit button pops up ui element to add/modify/delete the values
- subsequent save on the page synchronizes the allowed values 

todo: enable "Save" button on edit of the allowed values, tests, work out if sync behavior is what's wanted.

## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [ ] Risk label is set correctly
- [ ] All migrations are backwards compatible and won't block deploy
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [ ] If QA is part of the safety story, the "Awaiting QA" label is used
- [ ] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

kmtracey committed:
Beginnings of UI for specifying allowed values...

...for select data types in data dictionary.

- add model for storing the allowed values and their descriptions
- fetch existing allowed values when initial page data is fetched
- display allowed values and an edit button when appropriate
- next step: make edit button functional

orangejenny commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -55,6 +55,11 @@ hqDefine("data_dictionary/js/data_dictionary", [
         self.originalResourcePropPath = fhirResourcePropPath;
         self.deprecated = ko.observable(deprecated || false);
         self.removeFHIRResourcePropertyPath = ko.observable(removeFHIRResourcePropertyPath || false);
+        self.allowedValues = ko.observableArray();
```
Nitpick: this could be a little more concise by using `_.map` and setting `allowedValues` to the entire array instead of pushing individual values.

orangejenny commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -78,9 +83,30 @@ hqDefine("data_dictionary/js/data_dictionary", [
+            return self.dataType() == 'select';
+        };
+
+        self.allowedValuesDisplay = ko.computed(function() {
```
I'd make this a real list, using a `foreach` binding in the HTML and a `list-unstyled` class to avoid bullets, rather than br-separated text, both to be a tiny bit more semantic and to avoid needing to escape the values, which are an HTML injection risk.

orangejenny commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -78,9 +83,30 @@ hqDefine("data_dictionary/js/data_dictionary", [
             self.removeFHIRResourcePropertyPath(false);
         };
 
+        self.canHaveAllowedValues = function () {
```
This ought to be a `ko.computed` because it depends on an observable, even though `dataType` doesn't _really_ need to be an observable, since its value doesn't change.

orangejenny commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -78,9 +83,30 @@ hqDefine("data_dictionary/js/data_dictionary", [
+        self.description = description;
+
+        self.displayValue = function () {
+            var beginning = self.allowedValue ? self.allowedValue : '(empty)'
```
`gettext("empty")` will translate the text.

orangejenny reviewed (commented):
This looks good. My comments are minor (except the HTML injection one).

kmtracey committed:
Address initial review comments: try to be a bit more idiomatic/concise, and move html into template where it belongs.

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Use a uiElementKeyValueList to make the allowed values lists editable.

Additionally, change some text on the ui element (e.g. heading, add button) to be more user-friendly (reduce emphasis on "mapping"). Possibly these text bits should be made caller-configurable if these changes are not appropriate for all the places this widget is currently used.

Further, make the placeholders in the "mapping" fields caller-configurable.

Still TODO: enable "Save" button when the allowed values are edited, tests

orangejenny commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -100,6 +102,12 @@ <h3 data-bind="text: $root.activeCaseType()"></h3>
           </td>
+          <td>
+            <div data-bind="visible: canHaveAllowedValues()">
+              <b>{% trans "Allowed Values" %}</b>
```
I'd put this in the `<th>`, since this column is dedicated to allowed values.

orangejenny commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -155,6 +156,19 @@ def save_case_property(name, case_type, domain=None, data_type=None,
+            av_obj, _ = CasePropertyAllowedValue.objects.update_or_create(
+                case_property=prop, allowed_value=allowed_value, defaults={"description": av_desc})
+            obj_pks.append(av_obj.pk)
+        # Ddelete any database-resident allowed values that were not found in
```
```suggestion
        # Delete any database-resident allowed values that were not found in
```

orangejenny commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -155,6 +156,19 @@ def save_case_property(name, case_type, domain=None, data_type=None,
+        obj_pks = []
+        for allowed_value, av_desc in allowed_values.items():
+            av_obj, _ = CasePropertyAllowedValue.objects.update_or_create(
+                case_property=prop, allowed_value=allowed_value, defaults={"description": av_desc})
```
HQ should do some kind of validation on the allowed values. How about django's `validate_slug`? It's possible there will be projects that want more flexibility, but that seems like a decent starting point.

orangejenny commented on corehq/apps/hqwebapp/static/hqwebapp/js/ui_elements/ui-element-key-val-list.js:
```diff
@@ -34,12 +35,12 @@ hqDefine('hqwebapp/js/ui_elements/ui-element-key-val-list', [
 
         $modalContent.prepend('<div class="modal-header"><a class="close" data-dismiss="modal">×</a><h4 class="modal-title">'
-            + django.gettext('Edit Mapping for ') + this.modal_title + '</h4></div>');
+            + django.gettext('Edit ') + this.modal_title + '</h4></div>');
```
Could you make "Edit" part of the modal title? Combining translated strings doesn't allow the translator to re-order the words if needed.

orangejenny reviewed (commented):
This looks good, my feedback is minor.

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Enable Save button on allowed value modal "Done" click.

kmtracey committed:
Swith to the `on('change', ...)` made available via...

the `KeyValList` call to `hqMain.eventize` to enable the Save button, rather than listening to the Done button click.

This will fire regardless of how the modal is closed (x in top or Done button), which seems correct since the top X still records whatever changes have been made in what is subsequently shown to the user.

Also make the allowed values modal work properly for a new dynamically added property created after page load.

kmtracey committed:
Add the basics of export/import of allowed values in data dictionary.

kmtracey committed:
Address PR feedback and change to "valid values"...

...for user-visible references to what we have been calling allowed values. (Much of the code still uses the "alloewd value" terminology, if wanted I can work to change the internals as well, but that's a bigger task than time probably allows today.)

kmtracey committed:
Merge branch 'master' into allowed_values_ui

snopoke commented on corehq/apps/data_dictionary/views.py:
```diff
@@ -384,9 +426,10 @@ def _process_bulk_upload(bulk_file, domain):
                         if fhir_resource_prop_path and not fhir_resource_type:
                             error = _('Could not find resource type for {}').format(case_type)
                     if not error:
+                        allowed_values = allowed_value_info[case_type][name]
```
this errors if the property does not exist in the dictionary: https://sentry.io/organizations/dimagi/issues/2607113040/?query=is%3Aunresolved

kmtracey replied:
Yes I noticed that too when looking into QA-3457 and plan to fix both issues.

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Changes to account for new SQL model.

- include new model in domain export code
- ensure new model is deleted when domain is deleted

kmtracey committed:
Give user specific feedback on save of a too-long valid value. Refs QA-3469

kmtracey committed:
Beef up error catching/reporting on data dictionary upload/import.

Catch when a case type is missing valid values, and let the user know instead of producing a 500 error
Notice when a case property cell is empty, and let the user know
Notice when a property is listed as having valid values but not found in the base case type information, and let user know.

Also added some basic unit tests of the data dictionary export/import. Import tests for these error situations, export is a bit more basic (so far just tests the right sheets are in the exported file), but it should be easy enough to add on as the data dictionary is worked on more.

Refs QA-3457

kmtracey committed:
Couple of simplifications from review comments.
